libs/ibl/src/CubemapSH.cpp

```diff
@@ -233,130 +246,123 @@ std::unique_ptr<float3[]> CubemapSH::computeSH(JobSystem& js, const Cubemap& cm,
+    // "Stupid Spherical Harmonics (SH)" by Peter-Pike Sloan
+    // The simply come for expanding the computation of each SH function.
+    constexpr float M_SQRT_PI = 1.7724538509f;
+    constexpr float M_SQRT_3  = 1.7320508076f;
+    constexpr float M_SQRT_5  = 2.2360679775f;
+    constexpr float M_SQRT_15 = 3.8729833462f;
+    constexpr float A[numCoefs] = {
+                  1.0f / (2.0f * M_SQRT_PI),    // 0  0
+            -M_SQRT_3  / (2.0f * M_SQRT_PI),    // 1 -1
+             M_SQRT_3  / (2.0f * M_SQRT_PI),    // 1  0
+            -M_SQRT_3  / (2.0f * M_SQRT_PI),    // 1  1
+             M_SQRT_15 / (2.0f * M_SQRT_PI),    // 2 -2
+            -M_SQRT_15 / (2.0f * M_SQRT_PI),    // 3 -1
+             M_SQRT_5  / (4.0f * M_SQRT_PI),    // 3  0
+            -M_SQRT_15 / (2.0f * M_SQRT_PI),    // 3  1
+             M_SQRT_15 / (4.0f * M_SQRT_PI)     // 3  2
     };
 
-    CubemapUtils::process<State>(const_cast<Cubemap&>(cm), js,
-            [&](State& state, size_t y, Cubemap::Face f, Cubemap::Texel const* data, size_t dim) {
-        for (size_t x=0 ; x<dim ; ++x, ++data) {
-
-            float3 s(cm.getDirectionFor(f, x, y));
-
-            // sample a color
-            float3 color(Cubemap::sampleAt(data));
-
-            // take solid angle into account
-            color *= CubemapUtils::solidAngle(dim, x, y);
-
-            state.SH[0] += color * A[0];
-            state.SH[1] += color * A[1] * s.y;
-            state.SH[2] += color * A[2] * s.z;
-            state.SH[3] += color * A[3] * s.x;
-            state.SH[4] += color * A[4] * s.y * s.x;
-            state.SH[5] += color * A[5] * s.y * s.z;
-            state.SH[6] += color * A[6] * (3 * s.z * s.z - 1);
-            state.SH[7] += color * A[7] * s.z * s.x;
-            state.SH[8] += color * A[8] * (s.x * s.x - s.y * s.y);
-        }
-    },
-    [&](State& state) {
-        for (size_t i=0 ; i<numCoefs ; i++) {
-            SH[i] += state.SH[i];
-        }
-    });
-
-    return SH;
+    // and pre-multiply the SH by the scaling factors
+    // precompute the scaling factor K
+    const std::vector<float> K = Ki(numBands);
+
+    // below we do several things:
+    // 1) we pre-multiply the SH by the scaling factors (saves math in the shader)
+    // 2) we pre-multiply by the lambertian diffuse 1/pi (saves math in the shader)
+    // 3) we pre-condition the SH so the reconstruction can be as below (saves math in the shader):
+    //    This is done by applying the A[i]/K[i] factor
+    //          c += sh[0];
+    //          c += sh[1] * s.y;
+    //          c += sh[2] * s.z;
+    //          c += sh[3] * s.x;
+    //          c += sh[4] * s.y * s.x;
+    //          c += sh[5] * s.y * s.z;
+    //          c += sh[6] * (3 * s.z * s.z - 1);
+    //          c += sh[7] * s.z * s.x;
+    //          c += sh[8] * (s.x * s.x - s.y * s.y);
+
+    for (size_t i = 0; i < numCoefs; i++) {
+        SH[i] *= K[i] * M_1_PI * (A[i] / K[i]);
```

Comment:
The `K[i]`s cancel each other out, remove them?

Reply:
It does, but I want to show what's happening. It's only a handful of muls, so it doesn't matter. Also, in theory A/K should be constexpr pre-calculated constants.

This is in fact why we made the mistake above. We applied `K[i] * M_1_PI` instead of `A[i] * M_1_PI` but the "scaling coefficients" are the `K[i]`, not the `A[i]`. The `A[i]` appear only because we're simplifying the expression for the shader.